File: src/lib/bug-board/filters.ts
import { boardNowMs, personName, SPRINT_DAYS } from "@/lib/bug-board/data";
import { toEpochMs } from "@/lib/bug-board/format";
import {
  PRIORITIES,
  SEVERITIES,
  type Bug,
  type DeveloperStatus,
  type Environment,
  type Priority,
  type Severity,
  type TesterStatus,
} from "@/lib/bug-board/types";

export interface BugFilters {
  query: string;
  /** Only used while the board is unscoped, i.e. showing every project. */
  projectIds: string[];
  /**
   * An explicit hand-off list, set by a shared link such as
   * `?bugs=BUG-1042,BUG-1051`. Empty means "no id restriction".
   */
  bugIds: string[];
  testerStatus: TesterStatus[];
  developerStatus: DeveloperStatus[];
  severity: Severity[];
  priority: Priority[];
  assigneeIds: string[];
  reporterIds: string[];
  modules: string[];
  sprints: string[];
  environments: Environment[];
  /** `YYYY-MM-DD`, inclusive, matched against the created date. */
  createdFrom: string;
  createdTo: string;
}

/** The facet keys, i.e. every filter whose value is a list of selections. */
export const ARRAY_FILTER_KEYS = [
  "projectIds",
  "testerStatus",
  "developerStatus",
  "severity",
  "priority",
  "assigneeIds",
  "reporterIds",
  "modules",
  "sprints",
  "environments",
] as const;
export type ArrayFilterKey = (typeof ARRAY_FILTER_KEYS)[number];

export const EMPTY_FILTERS: BugFilters = {
  query: "",
  projectIds: [],
  bugIds: [],
  testerStatus: [],
  developerStatus: [],
  severity: [],
  priority: [],
  assigneeIds: [],
  reporterIds: [],
  modules: [],
  sprints: [],
  environments: [],
  createdFrom: "",
  createdTo: "",
};

/** Facets only, so the search field can be shown separately. */
export function activeFilterCount(filters: BugFilters): number {
  return (
    filters.projectIds.length +
    (filters.bugIds.length > 0 ? 1 : 0) +
    filters.testerStatus.length +
    filters.developerStatus.length +
    filters.severity.length +
    filters.priority.length +
    filters.assigneeIds.length +
    filters.reporterIds.length +
    filters.modules.length +
    filters.sprints.length +
    filters.environments.length +
    (filters.createdFrom ? 1 : 0) +
    (filters.createdTo ? 1 : 0)
  );
}

function matchesQuery(bug: Bug, needle: string): boolean {
  if (!needle) return true;
  const haystack = [
    bug.id,
    bug.title,
    bug.module,
    bug.component,
    bug.sprint,
    bug.description,
    personName(bug.reporterId),
    bug.assigneeId ? personName(bug.assigneeId) : "unassigned",
    ...bug.labels,
  ]
    .join(" ")
    .toLowerCase();
  return haystack.includes(needle);
}

function includedIn<T>(selected: T[], value: T): boolean {
  return selected.length === 0 || selected.includes(value);
}

export function filterBugs(bugs: Bug[], filters: BugFilters): Bug[] {
  const needle = filters.query.trim().toLowerCase();
  const from = filters.createdFrom ? toEpochMs(filters.createdFrom) : null;
  // The end of the day, so a single-day range keeps its own bugs.
  const to = filters.createdTo
    ? toEpochMs(filters.createdTo) + 86_400_000 - 1
    : null;

  return bugs.filter((bug) => {
    if (filters.bugIds.length > 0 && !filters.bugIds.includes(bug.id)) {
      return false;
    }
    if (!matchesQuery(bug, needle)) return false;
    if (!includedIn(filters.projectIds, bug.projectId)) return false;
    if (!includedIn(filters.testerStatus, bug.testerStatus)) return false;
    if (!includedIn(filters.developerStatus, bug.developerStatus)) return false;
    if (!includedIn(filters.severity, bug.severity)) return false;
    if (!includedIn(filters.priority, bug.priority)) return false;
    if (!includedIn(filters.modules, bug.module)) return false;
    if (!includedIn(filters.sprints, bug.sprint)) return false;
    if (!includedIn(filters.environments, bug.environment)) return false;
    if (!includedIn(filters.reporterIds, bug.reporterId)) return false;
    if (
      filters.assigneeIds.length > 0 &&
      !filters.assigneeIds.includes(bug.assigneeId ?? "unassigned")
    ) {
      return false;
    }
    const createdAt = toEpochMs(bug.createdAt);
    if (from !== null && createdAt < from) return false;
    if (to !== null && createdAt > to) return false;
    return true;
  });
}

export const SORT_KEYS = [
  "id",
  "title",
  "module",
  "severity",
  "priority",
  "reporter",
  "testerStatus",
  "developerStatus",
  "assignee",
  "attachments",
  "createdAt",
  "updatedAt",
] as const;
export type SortKey = (typeof SORT_KEYS)[number];
export type SortDirection = "asc" | "desc";

export interface Sort {
  key: SortKey;
  direction: SortDirection;
}

export const DEFAULT_SORT: Sort = { key: "updatedAt", direction: "desc" };

const SEVERITY_RANK: Record<Severity, number> = Object.fromEntries(
  SEVERITIES.map((severity, index) => [severity, index])
) as Record<Severity, number>;

const PRIORITY_RANK: Record<Priority, number> = Object.fromEntries(
  PRIORITIES.map((priority, index) => [priority, index])
) as Record<Priority, number>;

const TESTER_RANK: Record<TesterStatus, number> = {
  open: 0,
  reopened: 1,
  "under-review": 2,
  verified: 3,
  closed: 4,
};

const DEVELOPER_RANK: Record<DeveloperStatus, number> = {
  blocked: 0,
  backlog: 1,
  assigned: 2,
  "in-progress": 3,
  "ready-for-qa": 4,
  fixed: 5,
};

/** Numbers sort numerically, everything else with a locale compare. */
function compare(a: Bug, b: Bug, key: SortKey): number {
  switch (key) {
    case "id":
      return a.id.localeCompare(b.id, undefined, { numeric: true });
    case "title":
      return a.title.localeCompare(b.title);
    case "module":
      return a.module.localeCompare(b.module);
    case "severity":
      return SEVERITY_RANK[a.severity] - SEVERITY_RANK[b.severity];
    case "priority":
      return PRIORITY_RANK[a.priority] - PRIORITY_RANK[b.priority];
    case "reporter":
      return personName(a.reporterId).localeCompare(personName(b.reporterId));
    case "testerStatus":
      return TESTER_RANK[a.testerStatus] - TESTER_RANK[b.testerStatus];
    case "developerStatus":
      return DEVELOPER_RANK[a.developerStatus] - DEVELOPER_RANK[b.developerStatus];
    case "assignee":
      return personName(a.assigneeId).localeCompare(personName(b.assigneeId));
    case "attachments":
      return a.attachments.length - b.attachments.length;
    case "createdAt":
      return toEpochMs(a.createdAt) - toEpochMs(b.createdAt);
    case "updatedAt":
      return toEpochMs(a.updatedAt) - toEpochMs(b.updatedAt);
  }
}

export function sortBugs(bugs: Bug[], sort: Sort): Bug[] {
  const factor = sort.direction === "asc" ? 1 : -1;
  return [...bugs].sort((a, b) => {
    const result = compare(a, b, sort.key);
    // Ties fall back to the id so pagination stays stable.
    return result !== 0
      ? result * factor
      : a.id.localeCompare(b.id, undefined, { numeric: true });
  });
}

export function isOpen(bug: Bug): boolean {
  return bug.testerStatus !== "closed";
}

/** The facet equivalent of {@link isOpen}, for filters and shared links. */
export const OPEN_TESTER_STATUSES: TesterStatus[] = [
  "open",
  "under-review",
  "verified",
  "reopened",
];

export interface SummaryCard {
  id: string;
  label: string;
  value: number;
  hint: string;
  /** Change over the last seven days; `null` when a delta is not meaningful. */
  delta: number | null;
  /** Whether a rising number is a good thing, used to pick the arrow. */
  risingIsGood?: boolean;
}

export function summarize(bugs: Bug[]): SummaryCard[] {
  const now = boardNowMs();
  const week = now - 7 * 86_400_000;
  const sprintStart = now - SPRINT_DAYS * 86_400_000;

  const createdThisWeek = bugs.filter(
    (bug) => toEpochMs(bug.createdAt) >= week
  ).length;
  const open = bugs.filter(isOpen);
  const closedThisSprint = bugs.filter(
    (bug) => bug.testerStatus === "closed" && toEpochMs(bug.updatedAt) >= sprintStart
  );
  const closedThisWeek = bugs.filter(
    (bug) => bug.testerStatus === "closed" && toEpochMs(bug.updatedAt) >= week
  ).length;
  const critical = open.filter((bug) => bug.severity === "critical");
  const inProgress = bugs.filter((bug) => bug.developerStatus === "in-progress");
  const readyForQa = bugs.filter(
    (bug) => bug.developerStatus === "ready-for-qa"
  );
  const blocked = bugs.filter((bug) => bug.developerStatus === "blocked").length;
  const unassigned = open.filter((bug) => !bug.assigneeId).length;
  const criticalInProduction = critical.filter(
    (bug) => bug.environment === "production"
  ).length;
  const oldestReadyForQa = readyForQa.reduce((oldest, bug) => {
    const days = Math.floor((now - toEpochMs(bug.updatedAt)) / 86_400_000);
    return Math.max(oldest, days);
  }, 0);
  const verifiedThisSprint = bugs.filter(
    (bug) =>
      bug.testerStatus === "verified" && toEpochMs(bug.updatedAt) >= sprintStart
  ).length;

  return [
    {
      id: "total",
      label: "Total Bugs",
      value: bugs.length,
      hint: `${createdThisWeek} reported in the last 7 days`,
      delta: createdThisWeek,
      risingIsGood: false,
    },
    {
      id: "open",
      label: "Open Bugs",
      value: open.length,
      hint: `${unassigned} still unassigned`,
      delta: createdThisWeek - closedThisWeek,
      risingIsGood: false,
    },
    {
      id: "critical",
      label: "Critical Bugs",
      value: critical.length,
      hint: `${criticalInProduction} on production`,
      delta: null,
    },
    {
      id: "in-progress",
      label: "Bugs In Progress",
      value: inProgress.length,
      hint: blocked > 0 ? `${blocked} blocked` : "Nothing blocked",
      delta: null,
    },
    {
      id: "ready-for-qa",
      label: "Ready for QA",
      value: readyForQa.length,
      hint:
        readyForQa.length > 0
          ? `Oldest waiting ${oldestReadyForQa}d`
          : "Queue is empty",
      delta: null,
    },
    {
      id: "closed",
      label: "Closed This Sprint",
      value: closedThisSprint.length,
      hint: `${verifiedThisSprint} verified, awaiting close`,
      delta: closedThisWeek,
      risingIsGood: true,
    },
  ];
}

File: backend/apps/bugs/filters.py
"""Server-side twin of `filterBugs` / `sortBugs` in src/lib/bug-board/filters.ts.

The board's sort order is not alphabetical: severity, priority and both status
columns sort by workflow rank (critical first, blocked first, ...). Those ranks
are rebuilt here as SQL CASE expressions so a paginated request returns the same
rows in the same order the in-memory board did.
"""

import django_filters as filters
from django.db.models import (
    Case,
    Count,
    F,
    IntegerField,
    Q,
    TextField,
    Value,
    When,
)
from django.db.models.functions import Cast

from .models import Bug, DeveloperStatus, Priority, Severity, TesterStatus

#: The board's sentinel for "no assignee" inside the assignee facet.
UNASSIGNED = "unassigned"


def _rank(field: str, order: list[str]):
    return Case(
        *[When(**{field: value}, then=Value(index)) for index, value in enumerate(order)],
        default=Value(len(order)),
        output_field=IntegerField(),
    )


SEVERITY_ORDER = [Severity.CRITICAL, Severity.HIGH, Severity.MEDIUM, Severity.LOW]
PRIORITY_ORDER = [Priority.P0, Priority.P1, Priority.P2, Priority.P3]
TESTER_ORDER = [
    TesterStatus.OPEN,
    TesterStatus.REOPENED,
    TesterStatus.UNDER_REVIEW,
    TesterStatus.VERIFIED,
    TesterStatus.CLOSED,
]
DEVELOPER_ORDER = [
    DeveloperStatus.BLOCKED,
    DeveloperStatus.BACKLOG,
    DeveloperStatus.ASSIGNED,
    DeveloperStatus.IN_PROGRESS,
    DeveloperStatus.READY_FOR_QA,
    DeveloperStatus.FIXED,
]

#: Sort key from the frontend -> the annotation/field the database orders by.
SORT_FIELDS = {
    "id": "number",
    "title": "title",
    "module": "module",
    "severity": "_severity_rank",
    "priority": "_priority_rank",
    "reporter": "reporter__name",
    "testerStatus": "_tester_rank",
    "developerStatus": "_developer_rank",
    "assignee": "assignee__name",
    "attachments": "_attachment_count",
    "createdAt": "created_at",
    "updatedAt": "updated_at",
}

DEFAULT_SORT = "updatedAt"
DEFAULT_DIRECTION = "desc"


def annotate_for_sorting(queryset):
    return queryset.annotate(
        _severity_rank=_rank("severity", SEVERITY_ORDER),
        _priority_rank=_rank("priority", PRIORITY_ORDER),
        _tester_rank=_rank("tester_status", TESTER_ORDER),
        _developer_rank=_rank("developer_status", DEVELOPER_ORDER),
        _attachment_count=Count("attachments", distinct=True),
    )


def apply_sort(queryset, sort: str | None, direction: str | None):
    field = SORT_FIELDS.get(sort or DEFAULT_SORT, SORT_FIELDS[DEFAULT_SORT])
    descending = (direction or DEFAULT_DIRECTION).lower() == "desc"
    expression = F(field).desc(nulls_last=True) if descending else F(field).asc(
        nulls_first=True
    )
    # Ties fall back to the bug number, so pagination stays stable.
    return queryset.order_by(expression, "number")


class CommaSeparatedFilter(filters.BaseInFilter, filters.CharFilter):
    """Accepts both `?severity=high,low` and repeated `?severity=high&severity=low`."""

    def filter(self, qs, value):
        if not value:
            return qs
        values = [item for item in value if item]
        if not values:
            return qs
        return qs.filter(**{f"{self.field_name}__in": values})


class BugFilterSet(filters.FilterSet):
    query = filters.CharFilter(method="filter_query")
    bugIds = CommaSeparatedFilter(field_name="key")
    projectIds = CommaSeparatedFilter(field_name="project_id")
    testerStatus = CommaSeparatedFilter(field_name="tester_status")
    developerStatus = CommaSeparatedFilter(field_name="developer_status")
    severity = CommaSeparatedFilter(field_name="severity")
    priority = CommaSeparatedFilter(field_name="priority")
    reporterIds = CommaSeparatedFilter(field_name="reporter__username")
    assigneeIds = filters.CharFilter(method="filter_assignees")
    modules = CommaSeparatedFilter(field_name="module")
    sprints = CommaSeparatedFilter(field_name="sprint")
    environments = CommaSeparatedFilter(field_name="environment")
    createdFrom = filters.DateFilter(field_name="created_at", lookup_expr="date__gte")
    # `date__lte` rather than a raw timestamp compare, so a single-day range
    # still includes the bugs filed that day.
    createdTo = filters.DateFilter(field_name="created_at", lookup_expr="date__lte")

    class Meta:
        model = Bug
        fields: list[str] = []

    def filter_query(self, queryset, name, value):
        needle = (value or "").strip()
        if not needle:
            return queryset
        # `labels` is a Postgres array, which has no substring lookup, so it is
        # cast to text to match the frontend's "join every field and search it".
        return queryset.annotate(
            _labels_text=Cast("labels", TextField())
        ).filter(
            Q(key__icontains=needle)
            | Q(title__icontains=needle)
            | Q(module__icontains=needle)
            | Q(component__icontains=needle)
            | Q(sprint__icontains=needle)
            | Q(description__icontains=needle)
            | Q(reporter__name__icontains=needle)
            | Q(assignee__name__icontains=needle)
            | Q(_labels_text__icontains=needle)
        )

    def filter_assignees(self, queryset, name, value):
        """`unassigned` is a selectable option in the facet, not an empty filter."""
        selected = [item for item in (value or "").split(",") if item]
        if not selected:
            return queryset
        usernames = [item for item in selected if item != UNASSIGNED]
        condition = Q(assignee__username__in=usernames) if usernames else Q()
        if UNASSIGNED in selected:
            condition = condition | Q(assignee__isnull=True)
        return queryset.filter(condition)
